### DISCUSSION TITLE
Logging Cache information in Production/Preview environments

### DISCUSSION THREAD

**Discussion by williamtaggart97** [vercel/next.js#71662] [App Router]
I am building an application where I use next cache options and the fetch API to cache some requests with the data cache. I am trying to do an investigation into which requests are being cached (certain routes have only a 30% hit rate and I want to see if I'm missing something in production -- it's a 75% hit rate locally)

I'm using the following snippet to get logs in development, but I don't know where to get this in production. 
```
logging: {
    fetches: {
      fullUrl: true,
    },
  },
```

If anyone can point me in the direction of where this config flag turns on logging in the source code, I can probably figure it out from there. Or if anyone just knows if there's a header/something else in the response object, I'd greatly appreciate!

**Comment by ivoba** [comment]
Looks like the logRequests function is only included in the dev-server:
https://github.com/vercel/next.js/blob/655b808c5b352350fe85c0fadc7b92ff251e675b/packages/next/src/server/dev/next-dev-server.ts#L469

So logging is probably not possible in the standard nextJs server.

You could write a custom server with a custom requestHandler and include the logging there, just as in dev-server.
https://github.com/vercel/next.js/blob/655b808c5b352350fe85c0fadc7b92ff251e675b/examples/custom-server/server.ts

I would also like to have more control on what to log in production without having to build some custom server.

**Comment by gergokee** [comment]
i suggest you implement opentelemetry and you can log wherever you want :) 
https://nextjs.org/docs/app/building-your-application/optimizing/open-telemetry
For example console too:

#### **`instrumentation.ts`**:
```
import { registerOTel } from '@vercel/otel';
import { ConsoleSpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base';

export async function register() {
	registerOTel({
		serviceName: 'your-project-name',
		spanProcessors: [new SimpleSpanProcessor(new ConsoleSpanExporter())],
	});
}
```

**Reply by ivoba** [reply]
Seems a bit too much for just logging fetch requests.

And you probably will still have to write your own custom requestHandler that logs the fetch request/response in production.

**Reply by williamtaggart97** [reply]
I agree that a full `otel` setup feels like overkill but I was honestly just asking with the hope that nextjs had a built-in setting that turned the pre-built logs on in production. Building my own custom solution is currently the best way forward if I want that functionality.
